src/Renderer/MenuRenderer.php

```diff
@@ -0,0 +1,147 @@
+<?php declare(strict_types=1);
+
+namespace Becklyn\Menu\Renderer;
+
+use Becklyn\Menu\Item\MenuItem;
+use Becklyn\Menu\Visitor\ItemVisitor;
+use Becklyn\Menu\Voter\VoterInterface;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Contracts\Service\ServiceSubscriberInterface;
+use Twig\Environment;
+
+class MenuRenderer implements ServiceSubscriberInterface
+{
+    /**
+     * @var ContainerInterface
+     */
+    private $locator;
+
+    /**
+     * @var ItemVisitor[]
+     */
+    private $visitors;
+
+
+    /**
+     * @var VoterInterface[]
+     */
+    private $voters;
+
+
+    /**
+     * @param ContainerInterface $locator
+     * @param ItemVisitor[]      $visitors
+     * @param VoterInterface[]   $voters
+     */
+    public function __construct (ContainerInterface $locator, iterable $visitors, iterable $voters)
+    {
+        $this->locator = $locator;
+        $this->visitors = $visitors;
+        $this->voters = $voters;
+    }
+
+
+    /**
+     * @param MenuItem $root
+     * @param array    $options
+     *
+     * @return string
+     */
+    public function render (MenuItem $root, array $options = []) : string
+    {
+        // don't modify the original
+        $root = clone $root;
+
+        // apply external visitors
+        // must be applied before voters, as they can generate new nodes
+        $this->applyVisitors($root);
+
+        // run voters
+        $this->applyVoters($root);
+
+        // resolve options
+        $template = $options["template"] ?? "@BecklynMenu/menu.html.twig";
+        unset($options["template"]);
+
+        $options = \array_replace([
+            "translationDomain" => false,
+            "currentClass" => "is-current",
+            "ancestorClass" => "is-current-ancestor",
+            "depth" => null,
+        ], $options);
+
+        // resolve the ancestors
+        $root->resolveTree(
+            $this->locator->get(RouterInterface::class),
+            $options
+        );
+
+        return $this->locator->get(Environment::class)->render($template, [
+            "options" => $options,
+            "root" => $root,
+        ]);
+    }
+
+
+    /**
+     * Applies the visitors to the item and all children.
+     *
+     * @param MenuItem $item
+     */
+    private function applyVisitors (MenuItem $item)
+    {
+        foreach ($this->visitors as $visitor)
+        {
+            $visitor->visit($item);
+        }
+
+        foreach ($item->getChildren() as $child)
+        {
+            $this->applyVisitors($child);
+        }
+    }
+
+
+    /**
+     * Applies the voters to the item and all children.
+     *
+     * @param MenuItem $item
+     */
+    private function applyVoters (MenuItem $item)
+    {
+        // only apply voters if the item isn't yet marked as "current" from the construction
```

Comment:
Is it a viable use-case to change the `current` flag from within a voter only? As in, remove the voted-on `$item`'s current flag and set it on another item?

Reply:
This is the case if you set the item to `current` when *building* it.
If you want to change it while voting, just don't set anything when building it.